src/ape/managers/project/types.py

```diff
@@ -122,19 +126,30 @@ def file_needs_compiling(source: Path) -> bool:
                 return checksum != cached_checksum.hash  # Contents changed
 
             # NOTE: Filter by checksum to only update what's needed
-            needs_compiling = list(filter(file_needs_compiling, sources))
+            needs_compiling = set(filter(source_ids_need_compiling, source_paths))
+
+            # NOTE: Add referring source_id imports for each source path
+            referenced_source_ids: List[str] = []
+
+            for item in needs_compiling:
+                source_id = str(get_relative_path(item, self.contracts_folder))
+                referenced_source_ids.extend(cached_source_references.get(source_id, []))
+
+            needs_compiling.update(
+                [self.contracts_folder.joinpath(Path(s)) for s in referenced_source_ids]
```

Comment:
We _might_ need to get these source IDs relative to the base path before appending it to the `needs_compiling` list.

Reply:
Updated [4313364](https://github.com/ApeWorX/ape/pull/730/commits/43133645529061a923609b7346759a5ad5685b0a)